lib/std/Build/Step/Options.zig

```diff
@@ -41,135 +41,341 @@ pub fn addOption(self: *Options, comptime T: type, name: []const u8, value: T) v
+            }
+
+            if (name != null) {
+                try out.writeAll(";\n");
+            } else {
+                try out.writeAll(",\n");
             }
+            return;
         },
         .Void,
         .Bool,
         .Int,
         .ComptimeInt,
         .Float,
         .Null,
-        => try out.print("{any}", .{val}),
+        => {
+            if (name != null) {
+                try out.print("pub const {}: {s} = {any};\n", .{ std.zig.fmtId(name.?), @typeName(T), value });
+            } else {
+                try out.print("{any},\n", .{value});
+            }
+            return;
+        },
+        .Enum => |info| {
+            try printEnum(self, out, T, info, indent);
+
+            if (name != null) {
+                try out.print("pub const {}: {s} = .{s};\n", .{
+                    std.zig.fmtId(name.?),
+                    std.zig.fmtId(@typeName(T)),
+                    std.zig.fmtId(@tagName(value)),
+                });
+            }
+            return;
+        },
+        .Struct => |info| {
+            try printStruct(self, out, T, info, indent);
+
+            if (name != null) {
+                try out.print("pub const {}: {s} = ", .{
+                    std.zig.fmtId(name.?),
+                    std.zig.fmtId(@typeName(T)),
+                });
+                try printStructValue(self, out, info, value, indent);
+            }
+            return;
+        },
         else => @compileError(std.fmt.comptimePrint("`{s}` are not yet supported as build options", .{@tagName(@typeInfo(T))})),
     }
 }
 
+fn printUserDefinedType(self: *Options, out: anytype, comptime T: type, indent: u8) !void {
+    switch (@typeInfo(T)) {
+        .Enum => |info| {
+            return try printEnum(self, out, T, info, indent);
+        },
+        .Struct => |info| {
+            return try printStruct(self, out, T, info, indent);
+        },
+        else => {},
+    }
+}
+
+fn printEnum(self: *Options, out: anytype, comptime T: type, comptime val: std.builtin.Type.Enum, indent: u8) !void {
+    const gop = try self.encountered_types.getOrPut(@typeName(T));
+    if (!gop.found_existing) {
+        try out.writeByteNTimes(' ', indent);
+        try out.print("pub const {} = enum ({s}) {{\n", .{ std.zig.fmtId(@typeName(T)), @typeName(val.tag_type) });
+
+        inline for (val.fields) |field| {
+            try out.writeByteNTimes(' ', indent);
+            try out.print("    {} = {d},\n", .{ std.zig.fmtId(field.name), field.value });
+        }
+
+        if (!val.is_exhaustive) {
+            try out.writeByteNTimes(' ', indent);
+            try out.writeAll("    _,\n");
+        }
+
+        try out.writeByteNTimes(' ', indent);
+        try out.writeAll("};\n");
+    }
+}
+
+fn printStruct(self: *Options, out: anytype, comptime T: type, comptime val: std.builtin.Type.Struct, indent: u8) !void {
+    const gop = try self.encountered_types.getOrPut(@typeName(T));
+    if (!gop.found_existing) {
```

Comment:
Invert these conditions to reduce nesting:
```suggestion
    if (gop.found_existing) return;
```